helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -85,103 +87,121 @@ public Object deserialize(byte[] data) throws ZkMarshallingError {
         return data;
       }
     });
-    _zkBaseDataAccessor = new ZkBaseDataAccessor(_zkClient);
-
-    // TODO: Optimize serialization with Jackson
-    // TODO: Or use a better binary serialization protocol
-    // TODO: Consider making this also binary
-    // TODO: Consider an async write for the metadata as well
-    _znRecordClient = SharedZkClientFactory.getInstance()
-        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddr));
-    _znRecordBaseDataAccessor = new ZkBaseDataAccessor<>(_znRecordClient);
-    _znRecordClient.setZkSerializer(new ZNRecordSerializer());
-
+    _zkBaseDataAccessor = new ZkBaseDataAccessor<>(_zkClient);
     _zkSerializer = new ZNRecordJacksonSerializer();
     _bucketSize = bucketSize;
-    _numVersions = numVersions;
+    _versionTTL = versionTTL;
   }
 
   /**
    * Constructor that uses a default bucket size.
    * @param zkAddr
    */
   public ZkBucketDataAccessor(String zkAddr) {
-    this(zkAddr, DEFAULT_BUCKET_SIZE, DEFAULT_NUM_VERSIONS);
+    this(zkAddr, DEFAULT_BUCKET_SIZE, DEFAULT_VERSION_TTL);
   }
 
   @Override
   public <T extends HelixProperty> boolean compressedBucketWrite(String path, T value)
       throws IOException {
-    // Take the ZNrecord and serialize it (get byte[])
+    final long timestamp = System.currentTimeMillis();
+
+    DataUpdater<byte[]> lastWriteVersionUpdater = dataInZk -> {
+      if (dataInZk == null || dataInZk.length == 0) {
+        // No last write version exists, so start with 0
+        return ("0_" + timestamp).getBytes();
```

Comment:
What's your idea behind using an "integer + timestamp" combination pattern? Since timestamp itself is monotonically Increasing, and it's already used as Txid or version id. Can we get rid of the combination pattern, to just make things simpler?

Reply:
Also, I'm interesting to know why two different paths are needed: last write key + last successful write key. Can we simply just keep the "last successful key"? For example, remove the 1st step of asking for a version, directly use the timestamp as the version id. Once all steps succeed, overwrite the last successful key with the timestamp

Reply:
Please see https://github.com/apache/helix/wiki/Concurrency-and-Parallelism-for-BucketDataAccessor for details.
Timestamp cannot be used for monotonically-increasing sequences because of clock drift.

Reply:
Can you explain how the clock drift could happen in the helix controller, in the same data-center? We're not writing across data-centers, aren't we? e.g UTC time.

Reply:
It's possible if we consider cases where there could be multiple writers. Of course, in the case where there's one writer, using timestamp would work and would simplify things. But the purpose of this change is to increase parallelism, and that means we want to allow multiple writers (whether they are in the same data centers or not) as well.

Even if we had multiple writers in the same data center, NTP usually doesn't guarantee strict ordering of timestamps either.

Reply:
I'm concerned about the introduced first round trip of request/response, think about the latency added with a trade-off with the minor cases of timestamp not strictly increasing (I believe it happens rarely and many docs claim the usage of a simple timestamp in MVCC). It won't reduce the ability of multi-writers but will make the design simpler and performance better, I don't think the point is discussed well given our use cases.